### ISSUE TITLE
Tracking issue: implicit role cleanup

### ISSUE THREAD

**Issue by keithamus** [primer/react#3845]
### Description

A lot of our React components apply `role=` for elements which have an implicit role. In a lot of these cases, the `role=` assigned is that of the implicit role, in other words the attribute is entirely redundant. 

We can safely remove these roles that are explicitly set, and rely on the implicit role as defined by the HTML spec.

### Steps to reproduce

This effects a variety of components and so there are many reproductions. Visit some components which assign a role, compare the tag/role to the implicit role as defined in the specs.

For example:

`Table.tsx` uses `StyledTable` with `role=table`:

https://github.com/primer/react/blob/216d2a9f57b8acb0701ab4e04a23e057fc325c90/src/DataTable/Table.tsx#L279

However `StyledTable` is itself a `<table>` and so the role is redundant:
 https://github.com/primer/react/blob/216d2a9f57b8acb0701ab4e04a23e057fc325c90/src/DataTable/Table.tsx#L24

A more direct example; `<TableHead>` is a `<thead>` with `role=rowgroup`, but that's also redundant: https://github.com/primer/react/blob/216d2a9f57b8acb0701ab4e04a23e057fc325c90/src/DataTable/Table.tsx#L298

### Version

v35.32.1

### Browser

Chrome, Safari, Firefox, Edge, iOS Safari

**Comment by keithamus** [comment]
https://github.com/primer/react/pull/2996/commits/2c21c81cedf8a58fc7b409485db67fc1ebbdf502 added the roles to DataTable, the comment on some lines:

```js
    // We need to explicitly pass this role because some ATs and browsers drop table semantics
    // when we use `display: contents` or `display: grid` in the table
    // eslint-disable-next-line jsx-a11y/no-redundant-roles
```

**Comment by lesliecdubs** [comment]
@keithamus do you think this issue remains valid and should land in the Primer teams backlog? Asking in part because [continued discussion about these attributes](https://github.slack.com/archives/C01L618AEP9/p1697809185967159?thread_ts=1697794152.996809&cid=C01L618AEP9) uncovered their usefulness in ensuring compatibility across various screen reader/OS/browser combinations.

**Comment by keithamus** [comment]
Sure, let's close it.
